pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -293,7 +306,20 @@ func EndpointsEqualBeyondHash(ep1, ep2 *discovery.Endpoint) bool {
 		return false
 	}
 
-	if objectRefPtrChanged(ep1.TargetRef, ep2.TargetRef) {
+	// Serving and Terminating will only be set when the EndpointSliceTerminatingCondition feature is on.
+	// Ignore their difference if the expected or actual value is nil, which means the feature enablement is changed.
+	// Otherwise all EndpointSlices in the system would be updated on the first controller-manager restart even without
+	// actual changes, leading to delay in processing legitimate updates.
+	// Its value will be set to the expected one when there is an actual change triggering update of this EndpointSlice.
+	if ep1.Conditions.Serving != nil && ep2.Conditions.Serving != nil && *ep1.Conditions.Serving != *ep2.Conditions.Serving {
```

Comment:
This doesn't seem related or correct?

When Serving changes from nil to non-nil, this won't fire.  I think you want:
```
func ptrBoolEqual(a, b *bool) bool {
    if (a == nil) != (b == nil) {
        return false
    }
    if a == nil {
        return true
    }
    return *a == *b
}
```

and 

```
if ptrBoolEqual(ep1.Conditions.Serving, ep2.Conditions.Serving) {
```

Reply:
> When Serving changes from nil to non-nil, this won't fire

I think that's intentional, otherwise the first controller restart with the feature enabled instantly writes all endpoints?